File: app/src/main/java/com/niceplaces/niceplaces/activities/IntroActivity.java
package com.niceplaces.niceplaces.activities;

import android.content.Intent;
import android.os.Bundle;
import android.support.v4.view.ViewPager;
import android.support.v7.app.AppCompatActivity;
import android.text.Html;
import android.view.View;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.niceplaces.niceplaces.R;
import com.niceplaces.niceplaces.adapters.ViewPagerAdapter;

public class IntroActivity extends AppCompatActivity {

    private ViewPager mViewPager;
    private TextView[] mDots;
    private LinearLayout mLayoutDots;
    private Button mButtonNext, mButtonSkip;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final IntroActivity thisActivity = this;
        setContentView(R.layout.activity_intro);
        getSupportActionBar().hide();
        final int[] layouts = new int[]{
                R.layout.activity_intro1,
                R.layout.activity_intro2,
                R.layout.activity_intro3,
                R.layout.activity_intro4,
                R.layout.activity_intro5,
                R.layout.activity_intro6,
                R.layout.activity_intro7,
        };
        mLayoutDots = findViewById(R.id.layout_dots);
        mViewPager = findViewById(R.id.view_pager);
        mButtonNext = findViewById(R.id.button_next);
        mButtonSkip = findViewById(R.id.button_skip);
        ViewPagerAdapter adapter = new ViewPagerAdapter(this, layouts);
        mViewPager.setAdapter(adapter);
        addBottomDots(0, layouts.length);
        mViewPager.setOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {

            }

            @Override
            public void onPageSelected(int position) {
                addBottomDots(position, layouts.length);
                if (position == layouts.length - 1){
                    mButtonNext.setText("Vai!");
                    mButtonSkip.setVisibility(View.GONE);
                } else {
                    mButtonNext.setText("Avanti");
                    mButtonSkip.setVisibility(View.VISIBLE);
                }
            }

            @Override
            public void onPageScrollStateChanged(int state) {

            }
        });
        mButtonSkip.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent i = new Intent(thisActivity, MenuActivity.class);
                startActivity(i);
                finish();
            }
        });
        mButtonNext.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                int next = mViewPager.getCurrentItem() + 1;
                if (next < layouts.length){
                    mViewPager.setCurrentItem(next);
                } else {
                    Intent i = new Intent(thisActivity, MenuActivity.class);
                    startActivity(i);
                    finish();
                }
            }
        });
    }

    public void addBottomDots(int position, int number){
        mDots = new TextView[number];
        mLayoutDots.removeAllViews();
        for (int i = 0; i < mDots.length; i++){
            mDots[i] = new TextView(this);
            mDots[i].setText(Html.fromHtml("&#8226;"));
            mDots[i].setTextSize(35);
            if (i == position){
                mDots[i].setTextColor(getResources().getColor(R.color.colorPrimary));
            } else {
                mDots[i].setTextColor(getResources().getColor(R.color.colorPrimaryAlpha));
            }
            mLayoutDots.addView(mDots[i]);
        }
    }

}

File: app/src/main/java/com/niceplaces/niceplaces/activities/IntroActivity.kt
package com.niceplaces.niceplaces.activities

import android.content.Intent
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.viewpager.widget.ViewPager
import androidx.viewpager.widget.ViewPager.OnPageChangeListener
import com.niceplaces.niceplaces.R
import com.niceplaces.niceplaces.adapters.ViewPagerAdapter
import com.niceplaces.niceplaces.controllers.PrefsController
import com.niceplaces.niceplaces.utils.AppUtils

class IntroActivity : AppCompatActivity() {
    private lateinit var mViewPager: ViewPager
    private lateinit var mDots: Array<TextView?>
    private var mLayoutDots: LinearLayout? = null
    private lateinit var mButtonNext: Button
    private lateinit var mButtonSkip: Button
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val thisActivity = this
        setContentView(R.layout.activity_intro)
        supportActionBar!!.hide()
        val prefs = PrefsController(this)
        var layouts = intArrayOf(
                R.layout.activity_intro1,
                R.layout.activity_intro2,
                R.layout.activity_intro3,
                R.layout.activity_intro4,
                R.layout.activity_intro5,
                R.layout.activity_intro6,
                R.layout.activity_intro7,
//                R.layout.activity_intro_update,
                R.layout.activity_intro8)
        if (!prefs.isFistOpenAfterInstall && prefs.isFistOpenAfterUpdate){
            layouts = intArrayOf(R.layout.activity_intro_update)
        }
        prefs.firstOpenDone()
        mLayoutDots = findViewById(R.id.layout_dots)
        mViewPager = findViewById(R.id.view_pager)
        mButtonNext = findViewById(R.id.button_next)
        mButtonSkip = findViewById(R.id.button_skip)
        val adapter = ViewPagerAdapter(this, layouts)
        mViewPager.adapter = adapter
        addBottomDots(0, layouts.size)
        if (layouts.size == 1){
            mButtonNext.setText(R.string.go)
            mButtonSkip.visibility = View.GONE
        }
        mViewPager.addOnPageChangeListener(object : OnPageChangeListener {

            override fun onPageScrolled(position: Int, positionOffset: Float, positionOffsetPixels: Int) {}

            override fun onPageSelected(position: Int) {
                addBottomDots(position, layouts.size)
                if (position == layouts.size - 1) {
                    mButtonNext.setText(R.string.go)
                    mButtonSkip.visibility = View.GONE
                } else {
                    mButtonNext.setText(R.string.next)
                    mButtonSkip.visibility = View.VISIBLE
                }
            }

            override fun onPageScrollStateChanged(state: Int) {}
        })
        mButtonSkip.setOnClickListener {
            if (prefs.isPrivacyAccepted) {
                val i = Intent(thisActivity, MenuActivity::class.java)
                startActivity(i)
            } else {
                val i = Intent(thisActivity, PrivacyActivity::class.java)
                startActivity(i)
            }
            thisActivity.finish()
        }
        mButtonNext.setOnClickListener {
            val next = mViewPager.currentItem + 1
            if (next < layouts.size) {
                mViewPager.currentItem = next
            } else if (prefs.isPrivacyAccepted) {
                val i = Intent(thisActivity, MenuActivity::class.java)
                startActivity(i)
                thisActivity.finish()
            } else {
                val i = Intent(thisActivity, PrivacyActivity::class.java)
                startActivity(i)
                thisActivity.finish()
            }
        }
    }

    fun addBottomDots(position: Int, number: Int) {
        mDots = arrayOfNulls(number)
        mLayoutDots!!.removeAllViews()
        for (i in mDots.indices) {
            mDots[i] = TextView(this)
            mDots[i]!!.text = AppUtils.fromHtml("&#8226;")
            mDots[i]!!.textSize = 35f
            if (i == position) {
                mDots[i]!!.setTextColor(ContextCompat.getColor(this, R.color.colorPrimary))
            } else {
                mDots[i]!!.setTextColor(ContextCompat.getColor(this, R.color.colorPrimaryAlpha2))
            }
            mLayoutDots!!.addView(mDots[i])
        }
    }
}